src/rngs/os.rs

```diff
@@ -483,6 +488,127 @@ mod imp {
     }
 }
 
+// Read from `/dev/random`, with chunks of limited size (1040 bytes).
+// `/dev/random` uses the Hash_DRBG with SHA512 algorithm from NIST SP 800-90A.
+// `/dev/urandom` uses the FIPS 186-2 algorithm, which is considered less
+// secure. We choose to read from `/dev/random`.
+//
+// Since Solaris 11.3 the `getrandom` syscall is available. To make sure we can
+// compile on both Solaris and on OpenSolaris derivatives, that do not have the
+// function, we do a direct syscall instead of calling a library function.
+//
+// We have no way to differentiate between Solaris, illumos, SmartOS, etc.
+#[cfg(target_os = "solaris")]
+mod imp {
+    extern crate libc;
+    use {Error, ErrorKind};
+    use std::io;
+    use super::unix::*;
+
+    #[derive(Clone, Debug)]
+    pub struct OsRng(OsRngMethod);
+
+    #[derive(Clone, Debug)]
+    enum OsRngMethod {
+        GetRandom,
+        RandomDevice,
+    }
+
+    impl OsRng {
+        pub fn new() -> Result<OsRng, Error> {
+            if is_getrandom_available() {
+                return Ok(OsRng(OsRngMethod::GetRandom));
+            }
+
+            open_random_device("/dev/random", false)?;
+            Ok(OsRng(OsRngMethod::RandomDevice))
+        }
+
+        pub fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), Error> {
+            for slice in dest.chunks_mut(1040) {
+                match self.0 {
+                    OsRngMethod::GetRandom => getrandom_try_fill(slice)?,
+                    OsRngMethod::RandomDevice => read_random_device(slice)?,
+                 }
+            }
+            Ok(())
+        }
+    }
+
+    fn getrandom(buf: &mut [u8]) -> libc::c_long {
+        extern "C" {
+            fn syscall(number: libc::c_long, ...) -> libc::c_long;
+        }
+
+        const SYS_GETRANDOM: libc::c_long = 143;
+        const GRND_NONBLOCK: libc::c_uint = 0x0001;
+        const GRND_RANDOM: libc::c_uint = 0x0002;
+
+        unsafe {
+            syscall(SYS_GETRANDOM, buf.as_mut_ptr(), buf.len(),
+                    GRND_NONBLOCK | GRND_RANDOM)
+        }
+    }
+
+    fn getrandom_try_fill(dest: &mut [u8]) -> Result<(), Error> {
+        trace!("OsRng: reading {} bytes via getrandom", dest.len());
+        let mut read = 0;
+        let len = dest.len();
+        while read < len {
+            let result = getrandom(&mut dest[read..]);
+            if result == -1 {
+                let err = io::Error::last_os_error();
+                let kind = err.kind();
+                if kind == io::ErrorKind::Interrupted {
+                    continue;
+                } else if kind == io::ErrorKind::WouldBlock {
+                    // Potentially this would waste bytes, but since we use
+                    // /dev/urandom blocking only happens if not initialised.
+                    // Also, wasting the bytes in dest doesn't matter very much.
```

Comment:
Did you understand the [man page](https://docs.oracle.com/cd/E86824_01/html/E54765/getrandom-2.html)? Apparently the function returns -1 when it would block, and may set 0 or the number of bytes requested, and one should check the return value — but this is -1!

Reply:
That is an old comment, I noticed it after making the commit. I was going over the documentation a bit more right now.

Reply:
Hmm, the manual is not very clear. Both -1 and 0 can indicate an error. We should ask for bytes in chunks of 1024, while it are chunks of 1040 for `/dev/random`. And apparently it is guaranteed the chunk it either not or completely filled.

Reply:
1040 is an unusual chunk size. Yeah, I love confusing manuals (not).